src/librustc_mir/interpret/eval_context.rs

```diff
@@ -349,34 +314,39 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         }
     }
 
-    pub(super) fn monomorphize<T: TypeFoldable<'tcx> + Subst<'tcx>>(
+    /// Call this on things you got out of the MIR (so it is as generic as the current
+    /// stack frame), to bring it into the proper environment for this interpreter.
+    pub(super) fn subst_from_frame_and_normalize_erasing_regions<T: TypeFoldable<'tcx>>(
         &self,
-        t: T,
-    ) -> InterpResult<'tcx, T> {
+        value: T,
+    ) -> T {
         match self.stack.last() {
-            Some(frame) => Ok(self.monomorphize_with_substs(t, frame.instance.substs)?),
-            None => if t.needs_subst() {
-                throw_inval!(TooGeneric)
-            } else {
-                Ok(t)
-            },
+            Some(frame) => self.tcx.normalize_erasing_regions(
+                self.param_env,
+                value.subst(*self.tcx, frame.instance.substs),
+            ),
```

Comment:
This could use `tcx.subst_and_normalize_erasing_regions`, right?

Reply:
I intentionally inlined that function to avoid its assert.

Reply:
I [don't see an assert](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc/traits/codegen/mod.rs.html#80-100)?

Reply:
I must've gotten confused, looks like it's `normalize_erasing_late_bound_regions` that's the problematic one.